src/server.cpp

```diff
@@ -37,44 +35,45 @@ void Server::handle_receive(const error_code& error, std::size_t bytes)
         return;
     }
 
-    size_t length = recv_buffer_[2] * 256 + recv_buffer_[3];
+    try
+    {
+        Packet packet = makeResponse(Packet(m_recvBuffer, bytes));
+
+        m_socket.async_send_to(boost::asio::buffer(packet.makeSendBuffer("secret"), 20),
+            m_remoteEndpoint, std::bind(&Server::handleSend, this, pls::_1, pls::_2));
 
-    if (bytes < length)
+        startReceive();
+    }
+    catch (size_t length)
     {
         std::cout << "Error: request length " << bytes << "is less than specified in the request - " << length << "\n";
         return;
     }
+}
 
-    if (recv_buffer_[0] == ACCESS_REQUEST)
-        send_buffer_[0] = ACCESS_ACCEPT;
-    else
-        send_buffer_[0] = ACCESS_REJECT;
-
-    send_buffer_[1] = recv_buffer_[1];
-    send_buffer_[2] = 0;
-    send_buffer_[3] = 20;
-
-    for (size_t i = 0; i < 16; ++i)
-        send_buffer_[i + 4] = recv_buffer_[i + 4];
-
-    std::string secr("secret");
-
-    for (size_t i = 0; i < 6; ++i)
-        send_buffer_[i + 20] = secr[i];
+void Server::handleSend(const error_code& /*error*/, std::size_t /*bytes_transferred*/)
+{
+}
 
-    std::array<uint8_t, 16> md;
+const Packet Server::makeResponse(const Packet& packet)
+{
+    uint8_t type = packet.type();
 
-    MD5(send_buffer_.data(), 20 + secr.length(), md.data());
+    if (type == ACCESS_REQUEST)
+    {
+        type = ACCESS_ACCEPT;
+        std::cout << "Packet type: ACCESS_ACCEPT\n";
+    }
+    else
+    {
+        type = ACCESS_REJECT;
+        std::cout << "Packet type: ACCESS_REJECT\n";
+    }
 
-    for (size_t i = 0; i < 16; ++i)
-        send_buffer_[i + 4] = md[i];
+    uint8_t id = packet.id();
 
-    socket_.async_send_to(boost::asio::buffer(send_buffer_, 20), remote_endpoint_,
-        std::bind(&Server::handle_send, this, pls::_1, pls::_2));
+    std::array<uint8_t, 16> auth = packet.auth();
 
-    start_receive();
+    return Packet(type, id, auth);
```

Comment:
A lot of things are going on in this function. And many of them are not necessary. Consider the following:
```c++
Packet Server::makeResponse(const Packet& request)
{
    if (request.type() == ACCESS_REQUEST)
    {
        std::cout << "Packet type: ACCESS_ACCEPT\n";
        return Packet(ACCESS_ACCEPT, request.id(), request.auth());
    }
    std::cout << "Packet type: ACCESS_REJECT\n";
    return Packet(ACCESS_REJECT, request.id(), request.auth());
}
```

What did I change? First, I removed `const` from the return type. There is no reason for const when you returning a value (but a lot of reason if you returning a referemce). Second, I renamed `packet` to `request`. The word 'packet' does not tell you much, it's just a packet, and you know it already from its type. But 'request' tells you that the data in this valiable is a request. And it's type says that the request is of type 'Packet'. Which is twice as much information! Third, I removed all those useless intermediate variables. As I often say, you don't need variables unless it is absolutely necessary.